R/drop_file_ops.R

```diff
@@ -1,131 +1,254 @@
+#' @param allow_ownership_transfer Allow moves by owner even if it would result
+#'   in an ownership transfer for the content being moved. This does not apply
+#'   to copies. The default for this field is False.
 #' @export
 #' @examples \dontrun{
 #' write.csv(mtcars, file = "mt.csv")
 #' drop_upload("mt.csv")
 #' drop_create("drop_test2")
 #' drop_copy("mt.csv", "drop_test2/mt2.csv")
 #' }
-drop_copy <- function(from_path = NULL, to_path = NULL, root = "auto", verbose = FALSE, dtoken = get_dropbox_token())  {
-  move_url <- "https://api.dropbox.com/1/fileops/copy"
-     args <- as.list(drop_compact(c(root = root,
-                                    from_path = from_path,
-                                    to_path = to_path)))
-  # TODO
-  # here if to_path is just a path, append filename at the end
-  if(drop_exists(from_path)) {
-    # copy
-    x <-POST(move_url, config(token = dtoken), query = args, encode = "form")
-    res <- content(x)
-    if(!verbose) {
-      message(sprintf("%s copied to %s", basename(res$path), dirname(res$path)))
+drop_copy <-
+  function(from_path = NULL,
+           to_path = NULL,
+           allow_shared_folder = FALSE,
+           autorename = FALSE,
+           allow_ownership_transfer = FALSE,
+           verbose = FALSE,
+           dtoken = get_dropbox_token())  {
+    copy_url <- "https://api.dropboxapi.com/2/files/copy_v2"
+
+    from_path <- add_slashes(from_path)
+    to_path <- add_slashes(to_path)
+
+    # Copying a file into a folder
+    file_to_folder <-
+      c(drop_type(from_path) == "file",
+        drop_type(to_path) == "folder")
+    to_path <-
+      ifelse(all(file_to_folder), paste0(to_path, from_path), to_path)
+
+    # coping a folder to another folder
+    folder_to_folder <-
+      c(drop_type(from_path) == "folder",
+        drop_type(to_path) == "folder")
+    to_path <-
+      ifelse(all(folder_to_folder), paste0(to_path, from_path), to_path)
+
+    # Copying a file to a file
+    # Nothing to do, since both paths reflect origin and destination
+
+    # Copying a folder to an existing filename will result in a HTTP 409 (conflict error)
+
+    args <- drop_compact(
+      list(
+        from_path = from_path,
+        to_path = to_path,
+        allow_shared_folder = allow_shared_folder,
+        autorename = autorename,
+        allow_ownership_transfer = allow_ownership_transfer
+      )
+    )
+
+    if (drop_exists(from_path)) {
+      # copy
+      x <-
+        httr::POST(copy_url,
+                   httr::config(token = dtoken),
+                   body = args,
+                   encode = "json")
+      res <- httr::content(x)
+      if (!verbose) {
+        message(sprintf("%s copied to %s", from_path, res$metadata$path_lower))
+        invisible(res)
+      } else {
+        pretty_lists(res)
+        invisible(res)
+      }
     } else {
-      pretty_lists(res)
-      invisible(res)
+      stop("File or folder not found \n")
```

Comment:
do you want to error, or return FALSE? can't do both; `stop` prevents `FALSE` from being returned here.

Reply:
Removed `FALSE`